Source/CarthageKit/Xcode.swift

```diff
@@ -1542,15 +1563,25 @@ private func UUIDsFromDwarfdump(URL: NSURL) -> SignalProducer<Set<NSUUID>, Carth
 		}
 }
 
-/// Returns the URL of a binary inside a given framework.
-private func binaryURL(frameworkURL: NSURL) -> Result<NSURL, CarthageError> {
-	let bundle = NSBundle(path: frameworkURL.path!)
+/// Returns the URL of a binary inside a given package.
+private func binaryURL(packageURL: NSURL) -> Result<NSURL, CarthageError> {
+	let bundle = NSBundle(path: packageURL.path!)
+	let packageType = (bundle?.objectForInfoDictionaryKey("CFBundlePackageType") as? String).flatMap(PackageType.init)
 
-	if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
-		return .Success(frameworkURL.URLByAppendingPathComponent(binaryName))
-	} else {
-		return .Failure(.ReadFailed(frameworkURL, nil))
+	switch packageType {
+	case .Framework?:
+		if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
+			return .Success(packageURL.URLByAppendingPathComponent(binaryName))
+		}
+	case .dSYM?:
+		if let binaryName = packageURL.URLByDeletingPathExtension?.URLByDeletingPathExtension?.lastPathComponent {
+			let binaryURL = packageURL.URLByAppendingPathComponent("Contents/Resources/DWARF/\(binaryName)")
+			return .Success(binaryURL)
+		}
+	default: break
```

Comment:
My motivation around not returning from this case was that since both `case .Framework?` and `case .dSYM?` only return based on some conditional inside of this, they could both return a failure as well. To complete the control flow here, we'd need to either explicitly `return .Failure` inside of each case:

```swift
	switch packageType {
	case .Framework?:
		if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
			return .Success(packageURL.URLByAppendingPathComponent(binaryName))
		}

		return .Failure(.ReadFailed(packageURL, nil))
	case .dSYM?:
		if let binaryName = packageURL.URLByDeletingPathExtension?.URLByDeletingPathExtension?.lastPathComponent {
			let binaryURL = packageURL.URLByAppendingPathComponent("Contents/Resources/DWARF/\(binaryName)")
			return .Success(binaryURL)
		}

		return .Failure(.ReadFailed(packageURL, nil))
	case .None: return .Failure(.ReadFailed(packageURL, nil))
	}
```

or use `fallthrough` to get to the `default` case:

```swift
	switch packageType {
	case .Framework?:
		if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
			return .Success(packageURL.URLByAppendingPathComponent(binaryName))
		}

		fallthrough
	case .dSYM?:
		if let binaryName = packageURL.URLByDeletingPathExtension?.URLByDeletingPathExtension?.lastPathComponent {
			let binaryURL = packageURL.URLByAppendingPathComponent("Contents/Resources/DWARF/\(binaryName)")
			return .Success(binaryURL)
		}

		fallthrough
	case .None: return .Failure(.ReadFailed(packageURL, nil))
	}
``` 

or break without a return as we have here.

I'm cool with changing this to return in the `.None` here, but it won't let us delete the final return at the end of the function too unless we do a little bit of other tweaking.